enhancements/monitoring/replace-prometheus-adapter-with-metrics-server.md

```diff
@@ -0,0 +1,257 @@
+* 4.18 GA, enabled by default
+* 4.19 => Remove `prometheus-adapter` from CMO
+
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+More testing (upgrade, downgrade, scale).
+* Sufficient time for feedback including signals from telemetry about the customer adoption
+  * We will be adding a Gauge Metric named `cluster_monitoring_operator_metrics_server_enabled` in [cmo](https://github.com/openshift/cluster-monitoring-operator/blob/master/pkg/metrics/metrics.go) to measure this.
+* Counter-measures to avoid service degradation of the Kubelet.
+* Conduct load testing.
+
+#### Removing a deprecated feature
+
+Add notice in `4.15` release notes that `prometheus-adapter` is deprecated and should be repeated on every future releases.
+
+- In `4.19` remove the prometheus-adapter from CMO code base completely.
+
+### Upgrade / Downgrade Strategy
+
+TBD
+
+### Version Skew Strategy
+
+TBD
+
+### Operational Aspects of API Extensions
+
+`metrics-server` will be deployed in HA mode (2 replicas by default). We will define Pod AntiAffinity Rule in `metrics-server` Deployment spec and create `PodDisruptionBudget` config to meet HA criteria.
+
+`KubeAggregatedAPIErrors` and `KubeAggregatedAPIDown` alerts already exist to monitor the `APIService`. Additionaly metrics-server exposes [internal metrices](https://github.com/kubernetes-sigs/metrics-server/blob/796fc0f832c1ac444c44f88a952be87524456e07/pkg/server/metrics.go#L29-L46). A few of the metrices that can be useful:
+
+
+1. `tick_duration_seconds` : The total time spent collecting and storing metrics in seconds
+
+1. `request_duration_seconds` : Duration of requests to Kubelet API in seconds
+
+1. `request_total` : Number of requests sent to Kubelet API
+
+1. `points` : Number of metrics points stored
+
+#### Failure Modes
+
+N/A
+
+#### Support Procedures
+
+N/A
+
+## Implementation History
+
+N/A
+
+## Alternatives
+
+1. The alternative would be creating a standalone Metrics Server operator that would work in conjuntion with CMO.
+But for the foreseeable future we’ll have CMO in the picture, but with a reduced profile.
+So CMO deploying metrics-server and managing is a better option since CMO manages prometheus-adapter and 
+it will already have info to delete those while switching to metrics-server. Having independent operators
+could make migration trickier as we need a way for both operators to collaborate. It also avoids the overhead
+to maintain and manage another operator and we can utilize existing test framework available in CMO to do
+e2e testing as well.
+
+
+1. Another approach involves deploying the metrics-server through CRD-based configuration. Given that all other configurations are provided via the `openshift-monitoring/cluster-monitoring-config` Config Map, this seems to be a more organized approach for the sake of clarity.
+
```

Comment:
Another alternative could be to have metrics-server managed by CVO. As far as I can tell, this could especially make sense on SNO, but it would require a hard switch from prometheus-adapter.

Reply:
Isn't this equivalent to the first alternative (e.g. a dedicated metrics-server-operator)?

Reply:
Thinking about it again, yes it is fairly similar